File: Chauffeur/Deliverables/SettingsDeliverable.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Chauffeur.Host;

namespace Chauffeur.Deliverables
{
    [DeliverableName("settings")]
    public sealed class SettingsDeliverable : Deliverable
    {
        private readonly IChauffeurSettings settings;

        public SettingsDeliverable(TextReader reader, TextWriter writer, IChauffeurSettings settings) : base(reader, writer)
        {
            this.settings = settings;
        }

        public async override Task<DeliverableResponse> Run(string command, string[] args)
        {
            var dic = new Dictionary<string, string>
            {
                { "Umbraco Version", settings.UmbracoVersion },
                { "Chauffeur Version", settings.ChauffeurVersion }
            };

            if (settings.TryGetSiteRootDirectory(out string path))
                dic.Add("Site Root", path);
            else
                dic.Add("Site Root", "Failed to access");

            if (settings.TryGetUmbracoDirectory(out path))
                dic.Add("Umbraco Directory", path);
            else
                dic.Add("Umbraco Directory", "Failed to access");

            if (settings.TryGetChauffeurDirectory(out path))
                dic.Add("Chauffeur Directory", path);
            else
                dic.Add("Chauffeur Directory", "Failed to access");

            dic.Add("Connection String", settings.ConnectionString.ConnectionString);

            await Out.WriteTableAsync(dic.Keys.Select(key => new
            {
                Setting = key,
                Value = dic[key]
            }));

            return DeliverableResponse.Continue;
        }
    }
}

File: src/Chauffeur/Deliverables/SettingsDeliverable.fs
namespace Chauffeur.Deliverables

open Chauffeur
open System.Collections.Generic
open FSharp.Control.Tasks.V2

type Row = { Setting: string; Value: string }

[<DeliverableName("settings")>]
type SettingsDeliverable(reader, writer, settings : IChauffeurSettings) =
    inherit Deliverable(reader, writer)

    override __.Run _ _ =
        let dic = new Dictionary<string, string>()

        dic.Add("Umbraco Version", settings.UmbracoVersion)
        dic.Add("Chauffeur Version", settings.ChauffeurVersion)

        dic.Add("Site Root", match settings.TryGetSiteRootDirectory() with
                             | true, path -> path
                             | false, _ -> "Failed to access")

        dic.Add("Umbraco Directory", match settings.TryGetUmbracoDirectory() with
                                     | true, path -> path
                                     | false, _ -> "Failed to access")

        dic.Add("Chauffeur Directory", match settings.TryGetChauffeurDirectory() with
                                       | true, path -> path
                                       | false, _ -> "Failed to access")

        dic.Add("Connection String", settings.ConnectionString.ConnectionString)

        let rows = dic
                   |> Seq.map(fun d -> { Setting = d.Key; Value = dic.[d.Key] })

        task {
            do! TextWriterExtensions.WriteTableAsync writer rows (Set.empty |> dict)
            return DeliverableResponse.Continue
        }

